pkg/virtual/apiexport/authorizer/maximal_permission_policy.go

```diff
@@ -0,0 +1,257 @@
+		},
+		getAPIBindingReferenceForAttributes: func(attr authorizer.Attributes, clusterName logicalcluster.Name) (*apisv1alpha1.ExportReference, bool, error) {
+			return kcpauth.GetAPIBindingReferenceForAttributes(apiBindingIndexer, attr, clusterName)
+		}}
+}
+
+func (a *maximalPermissionAuthorizer) Authorize(ctx context.Context, attr authorizer.Attributes) (authorizer.Decision, string, error) {
+	lcluster, err := genericapirequest.ClusterNameFrom(ctx)
+	if err != nil {
+		kaudit.AddAuditAnnotations(
+			ctx,
+			VirtualAPIExportMaximumPermissionPolicyAuditDecision, kcpauth.DecisionNoOpinion,
+			VirtualAPIExportMaximumPermissionPolicyAuditReason, fmt.Sprintf("error getting cluster from request: %v", err),
+		)
+		return authorizer.DecisionNoOpinion, VirtualAPIExportNotPermittedReason, err
+	}
+
+	if lcluster != logicalcluster.Wildcard {
+		_, bound, err := a.getAPIBindingReferenceForAttributes(attr, lcluster)
+		if err != nil {
+			kaudit.AddAuditAnnotations(
+				ctx,
+				VirtualAPIExportMaximumPermissionPolicyAuditDecision, kcpauth.DecisionNoOpinion,
+				VirtualAPIExportMaximumPermissionPolicyAuditReason, fmt.Sprintf("error getting API binding reference: %v", err),
+			)
+			return authorizer.DecisionNoOpinion, VirtualAPIExportNotPermittedReason, err
+		}
+
+		if !bound {
+			kaudit.AddAuditAnnotations(
+				ctx,
+				VirtualAPIExportMaximumPermissionPolicyAuditDecision, kcpauth.DecisionAllowed,
+				VirtualAPIExportMaximumPermissionPolicyAuditReason, "no API binding bound",
+			)
+			return authorizer.DecisionAllow, VirtualAPIExportPermittedReason, nil
+		}
+	}
+
+	apiDomainKey := dynamiccontext.APIDomainKeyFrom(ctx)
+	parts := strings.Split(string(apiDomainKey), "/")
+	if len(parts) < 2 {
+		err := fmt.Errorf("invalid API domain key: %q", apiDomainKey)
+		kaudit.AddAuditAnnotations(
+			ctx,
+			VirtualAPIExportMaximumPermissionPolicyAuditDecision, kcpauth.DecisionNoOpinion,
+			VirtualAPIExportMaximumPermissionPolicyAuditReason, err.Error(),
+		)
+		return authorizer.DecisionNoOpinion, VirtualAPIExportNotPermittedReason, err
+	}
+
+	apiExportCluster := parts[0]
+	apiExportName := parts[1]
+
+	apiExport, err := a.getAPIExport(apiExportCluster, apiExportName)
+	if err != nil && kerrors.IsNotFound(err) {
+		kaudit.AddAuditAnnotations(
+			ctx,
+			VirtualAPIExportMaximumPermissionPolicyAuditDecision, kcpauth.DecisionNoOpinion,
+			VirtualAPIExportMaximumPermissionPolicyAuditReason, fmt.Sprintf("API export %q not found", apiDomainKey),
+		)
+		return authorizer.DecisionNoOpinion, VirtualAPIExportNotPermittedReason, err
+	}
+
+	identityHash := getPermissionClaimIdentity(apiExport, attr)
+	if identityHash == "" {
+		kaudit.AddAuditAnnotations(
+			ctx,
+			VirtualAPIExportMaximumPermissionPolicyAuditDecision, kcpauth.DecisionAllowed,
+			VirtualAPIExportMaximumPermissionPolicyAuditReason, "claimed resource does not have an identity hash",
+		)
+		return authorizer.DecisionAllow, notEnforcedMaximumPermissionPolicyReason, nil
```

Comment:
Why wouldn't we enforce maximal permission policy on resources without an identity hash? example would be a secret, right?

Reply:
Right, but we certainly want to have permission policies on those, no?

Reply:
according to my discussions with @sttts maximum permission policy cannot be applied in this case.

Reply:
Isn't the implication here then that an API provider that claims some k8s resource (say, a particular Secret or ConfigMap) could set a maxmial permission on that resource from the user? I'm confused about the use-cases here

Reply:
You have a point here 🤔 I believe we have to separete still the two request cases:

- _non-wildcard request for a concrete cluster_: in this case a unique API binding can be found and thus an API export whose maximal permission can be asserted, even for native types. This is currently missing here but present in the existing API binding authorizer.
- _wildcard requests_: In this case no uniqe API binding can be found but still we can try to determine an API export via the identity hash of the claimed resource, if it exists. In this case an API export can be found and asserted. However if it cannot be found, contrary to the current logic, we should deny access.

Let me code this down, and also present the different use cases in the community meeting, and also document those as it is indeed tricky.

Reply:
> Isn't the implication here then that an API provider that claims some k8s resource (say, a particular Secret or ConfigMap) could set a maxmial permission on that resource from the user?

afaik this is not possible, we cannot set maximal permission policies on k8s resources.

Reply:
You cannot set a maximal permission policy on i.e. a secret or configmap, only on exported types. To have more clarification we made this an explicit comment.